src/tree/hist/histogram.h

```diff
@@ -238,8 +241,76 @@ common::BlockedSpace2d ConstructHistSpace(Partitioner const &partitioners,
       k++;
     }
   }
+
+  // Estimate the size of each data block based on model parameters and L1 capacity
+  // The general idea is to keep as much working-set data in L1 as possible.
+  /* Each processed row occupies ~32 bytes in L1:
+   * - gradient pair (p_gpair): 2 * sizeof(float)
+   * - row index (rid[i]): sizeof(size_t)
+   * - icol_start and icol_end: 2 * sizeof(size_t)
+   */
+  std::size_t l1_row_foot_print = (2 * sizeof(float) + 3 * sizeof(size_t));
+  double usable_l1_size = 0.8 * l1_size;
+
+  std::size_t space_in_l1_for_rows;
+  if (read_by_column) {
+   /* In this case, an accurate block_size estimate is performance-critical.
+    * For column-wise histogram construction, each column is processed over the
+    * same block of rows. If the block fits in L1, the row data are loaded once
+    * and reused across all columns; otherwise, the cache must be refilled for
+    * each column.
+    */
+
+    /* First step: determine whether one histogram column fits into L1.
+    * The maximum number of elements in a column is 2^8, 2^16, or 2^32,
```

Comment:
Could you please elaborate on what it means to be the maximum number of elements in a (histogram) column? I thought that's the number of histogram bins?

Reply:
> You are right, it is better to use max_bin in this case, otherwise the estimation would be too conservative

I didn't make any suggestion? I was curious about the constraint and where the numbers `2^8, 2^16` originate or what they are for.